src/libm-tester/tester.c

```diff
@@ -124,3202 +113,4205 @@ uint64_t d2u(double d) {
 
 //
 
-boolean isPlusZero(double x) { return x == 0 && copysign(1, x) == 1; }
-boolean isMinusZero(double x) { return x == 0 && copysign(1, x) == -1; }
-boolean xisnan(double x) { return x != x; }
-double sign(double d) { return d < 0 ? -1 : 1; }
-
-boolean cmpDenorm(double x, double y) {
-  if (xisnan(x) && xisnan(y)) return true;
-  if (xisnan(x) || xisnan(y)) return false;
-  if (isinf(x) != isinf(y)) return false;
-  if (x == POSITIVE_INFINITY && y == POSITIVE_INFINITY) return true;
-  if (x == NEGATIVE_INFINITY && y == NEGATIVE_INFINITY) return true;
-  if (y == 0) {
-    if (isPlusZero(x) && isPlusZero(y)) return true;
-    if (isMinusZero(x) && isMinusZero(y)) return true;
-    return false;
+#define child_d_d(funcStr, arg) {				\
+    char str[256];						\
+    uint64_t u;							\
+    sprintf(str, funcStr " %" PRIx64 "\n", d2u(arg));		\
+    write(ptoc[1], str, strlen(str));				\
+    if (readln(ctop[0], str, 255) < 1) stop("child " funcStr);	\
+    sscanf(str, "%" PRIx64, &u);				\
+    return u2d(u);						\
   }
-  if (!xisnan(x) && !xisnan(y) && !isinf(x) && !isinf(y)) return sign(x) == sign(y);
-  return false;
-}
 
-long double ulp(long double x) {
-  x = fabsl(x);
-  int exp;
-
-  if (x == 0) {
-    return DENORMAL_DBL_MIN;
-  } else {
-    frexpl(x, &exp);
+#define child_d2_d(funcStr, arg) {				\
+    char str[256];						\
+    uint64_t u, v;						\
+    sprintf(str, funcStr " %" PRIx64 "\n", d2u(arg));		\
+    write(ptoc[1], str, strlen(str));				\
+    if (readln(ctop[0], str, 255) < 1) stop("child " funcStr);	\
+    sscanf(str, "%" PRIx64 " %" PRIx64, &u, &v);		\
+    Sleef_double2 ret;						\
+    ret.x = u2d(u);						\
+    ret.y = u2d(v);						\
+    return ret;							\
   }
 
-  return fmax(ldexp(1.0, exp-53), DENORMAL_DBL_MIN);
-}
-
-double countULP(long double x, long double y) {
-  double fx = x;
-  double fy = y;
-  if (xisnan(fx) && xisnan(fy)) return 0;
-  if (xisnan(fx) || xisnan(fy)) return 10000;
-  if (isinf(fx)) {
-    if (sign(fx) == sign(fy) && fabs(fy) > 1e+300) return 0; // Relaxed infinity handling
-    return 10001;
-  }
-  if (fx == POSITIVE_INFINITY && fy == POSITIVE_INFINITY) return 0;
-  if (fx == NEGATIVE_INFINITY && fy == NEGATIVE_INFINITY) return 0;
-  if (fy == 0) {
-    if (fx == 0) return 0;
-    return 10002;
-  }
-  if (!xisnan(fx) && !xisnan(fy) && !isinf(fx) && !isinf(fy)) {
-    return fabsl((x - y) / ulp(y));
+#define child_d_d_d(funcStr, arg1, arg2) {				\
+    char str[256];							\
+    uint64_t u;								\
+    sprintf(str, funcStr " %" PRIx64 " %" PRIx64 "\n", d2u(arg1), d2u(arg2)); \
+    write(ptoc[1], str, strlen(str));					\
+    if (readln(ctop[0], str, 255) < 1) stop("child " funcStr);		\
+    sscanf(str, "%" PRIx64, &u);					\
+    return u2d(u);							\
   }
-  return 10003;
-}
+
+double child_sin(double x) { child_d_d("sin", x); }
```

Comment:
This would be just a matter of preference, but if the function declaration is included in the macro, the resulting code looks quite different from the standard C syntax.